test-data/unit/check-dataclasses.test

```diff
@@ -1032,3 +1032,275 @@ class B(A):
+
+[typing fixtures/typing-full.pyi]
+[builtins fixtures/dict.pyi]
+
+[case testDataclassesAsdictDictSubclass]
+from dataclasses import dataclass, asdict
+from typing import Dict, Generic, TypeVar
+
+@dataclass
+class Person:
+    name: str
+    age: int
+
+_KT = TypeVar("_KT")
+_VT = TypeVar("_VT")
+_Other = TypeVar("_Other")
+class MyDict(Dict[_KT, _VT], Generic[_Other, _KT, _VT]):
+    pass
+
+@dataclass
+class Course:
+    participants_by_name: MyDict[int, str, Person]
+
+instance = Course(participants_by_name=MyDict[int, str, Person]([("Joe", Person("Joe", 32))]))
+result = asdict(instance)
+reveal_type(result['participants_by_name']) # N: Revealed type is 'builtins.dict[builtins.str*, TypedDict({'name'?: builtins.str, 'age'?: builtins.int})]'
+
+[typing fixtures/typing-full.pyi]
+[builtins fixtures/dict.pyi]
+
+
+[case testDataclassesAsdictTuple]
+from dataclasses import dataclass, asdict
+from typing import Tuple
+
+@dataclass
+class Person:
+    name: str
+    age: int
+
+@dataclass
+class Course:
+    partners: Tuple[Person, Person]
+
+instance = Course(partners=(Person("Joe", 32), Person("John", 23)))
+result = asdict(instance)
+reveal_type(result['partners']) # N: Revealed type is 'Tuple[TypedDict({'name'?: builtins.str, 'age'?: builtins.int}), TypedDict({'name'?: builtins.str, 'age'?: builtins.int})]'
+
+
+[typing fixtures/typing-full.pyi]
+[builtins fixtures/tuple.pyi]
+
+
+
+[case testDataclassesAsdictTupleSubclass]
+from dataclasses import dataclass, asdict
+from typing import Tuple, Generic, TypeVar
+
+@dataclass
+class Person:
+    name: str
+    age: int
+
+class MyTuple(Tuple[Person, str]):
+    pass
+
+@dataclass
+class Course:
+    tuple_subclass: MyTuple
+
+instance = Course(tuple_subclass=MyTuple())
+result = asdict(instance)
+
+# For now, subclasses of Tuple are transformed to the Tuple base class
+# This is because the subclass, if it itself contains dataclass fields, may be transformed in such a way that it
+# is no longer compatible with the original Tuple class it is extending.
+reveal_type(result['tuple_subclass']) # N: Revealed type is 'Tuple[TypedDict({'name'?: builtins.str, 'age'?: builtins.int}), builtins.str]'
+
+
+[typing fixtures/typing-full.pyi]
+[builtins fixtures/tuple.pyi]
+
+[case testDataclassesAsdictNamedTuple]
+from dataclasses import dataclass, asdict
+from typing import NamedTuple
+
+@dataclass
+class Person:
+    name: str
+    age: int
+
+
+class Staff(NamedTuple):
+    teacher: Person
+    assistant: Person
+
+    def staff_method(self):
+        pass
+
+@dataclass
+class Course:
+    staff: Staff
+
+instance = Course(staff=Staff(teacher=Person("Joe", 32), assistant=Person("John", 23)))
+result = asdict(instance)
+# Due to implementation limitations, namedtuples are transformed to Any
+reveal_type(result['staff']) # N: Revealed type is 'Any'
+
+[typing fixtures/typing-full.pyi]
+[builtins fixtures/tuple.pyi]
```

Comment:
I would add couple fine grained (daemon) test cases. The tricky part is that when type of a dataclass attribute changes, then the call to `asdict()` should be reprocessed (put it in a different module). I am not 100% sure we already have enough extra dependencies added by the plugin to guarantee this, so such tests will be really useful.

Reply:
I added one. It revealed that there was no dependency on the fallback type for TypedDict. I tried to add that. Since it depends on python version which TypedDict fallback to use (I think?), I had to add a dependency on all modules (e.g. typing/typing_extensions/mypy_extensions) to the default plugin. Doesn't seem very nice, but I'm not sure of the right way to do it. Do you have any ideas?

Do you have any ideas for more fine-grained tests to add?